apps/moderation/api/tests.py

```diff
@@ -0,0 +1,205 @@
+import json
+import random
+from typing import TYPE_CHECKING, Callable, Optional, Union
+
+import pytest
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import ForeignKey
+from django.urls import reverse
+from rest_framework.test import APIClient
+
+from apps.moderation.models import Change, ContentContribution, ModeratedModel
+from apps.users.models import User
+
+if TYPE_CHECKING:
+    from django.db.models.manager import Manager
+    from rest_framework.response import Response
+
+
+def shuffled_copy(data, size=None):
+    copy = data.copy()
+    random.shuffle(copy)
+    if size:
+        return copy[:size]
+    return copy
+
+
+@pytest.mark.django_db()
+@pytest.mark.usefixtures('api_client')
+class ModerationApiTest:
+    """Base moderation api test case."""
+
+    # Only inheriting classes should publish tests.
+    __test__ = False
+
+    api_client: APIClient
+
+    # api namespace, ex: quotes_api, entities_api
+    api_name: str
+    # api prefix, ex: quote, entity
+    api_prefix: str
+    # api path suffix for nested endpoints like Sources, ex: articles
+    # reverse urls will be rewritten to include this suffix if it doesn't contain already
+    api_path_suffix: str = None
+
+    contributor: User
+    # verified moderated model to be used update/patch/delete tests
+    verified_model: ModeratedModel
+    content_type: ContentType
+
+    # fields to be treated as relation fields
+    # TODO: could be improved to detect relation fields automatically via model._meta.get_field
+    relation_fields = []
+    # fields that won't be verified after creation/update/patch
+    # TODO: usually date fields are not checkable, find a way to check them or detect them automatically
+    uncheckable_fields = []
+    # test data to be used for creation and update/patch respectively
+    test_data: dict
+    updated_test_data: dict
+
+    def _test_api_view_get(self, view, url_kwargs=None, status_code=200):
+        path = reverse(f'{self.api_name}:{view}', kwargs=url_kwargs)
+        response = self.api_client.get(path)
+        assert response.status_code == status_code, 'Incorrect status code'
+
+    def _test_api_moderation_view(
+        self,
+        data: dict,
+        view: str = 'api-root',
+        url_kwargs=None,
+        change_status_code: int = 200,
+        method: str = 'post',
+        object_id: Optional[int] = None,
+    ):
+        if url_kwargs is None:
+            url_kwargs = {}
+        if object_id:
+            url_kwargs.update({'pk_or_slug': object_id})
+        path = reverse(f'{self.api_name}:{view}', kwargs=url_kwargs)
+        # Force request path to include API path suffix.
+        if self.api_path_suffix and self.api_path_suffix not in path:
+            path += self.api_path_suffix + '/'
```

Comment:
We might need to add this suffix for `_test_api_view_get` too. I'm guessing it's currently not failing because it's probably accessing parent path, i.e for articles, `/api/sources/` instead of `/api/sources/articles/`